util/src/trie/triedbmut.rs

```diff
@@ -259,259 +444,280 @@ impl<'db> TrieDBMut<'db> {
+
+					// always replace because whatever we get out here is not the branch we started with.
+					InsertAction::Replace(self.insert_inspector(branch, partial, value).unwrap_node())
+				} else if cp == existing_key.len() {
+				    trace!(target: "trie", "complete-prefix (cp={:?}): AUGMENT-AT-END", cp);
+
+					// fully-shared prefix for an extension.
+					// make a stub branch and an extension.
+					let branch = Node::Branch(empty_children(), Some(stored_value));
+					// augment the new branch.
+					let branch = self.insert_inspector(branch, partial.mid(cp), value).unwrap_node();
+
+					// always replace since we took a leaf and made an extension.
+					let branch_handle = self.storage.alloc(Stored::New(branch)).into();
+					InsertAction::Replace(Node::Extension(existing_key.encoded(false), branch_handle))
+				} else {
+					trace!(target: "trie", "partially-shared-prefix (exist={:?}; new={:?}; cp={:?}): AUGMENT-AT-END", existing_key.len(), partial.len(), cp);
+
+					// partially-shared prefix for an extension.
+					// start by making a leaf.
+					let low = Node::Leaf(existing_key.mid(cp).encoded(true), stored_value);
+					// augment it. this will result in the Leaf -> cp == 0 routine,
+					// which creates a branch.
+					let augmented_low = self.insert_inspector(low, partial.mid(cp), value).unwrap_node();
+
+					// make an extension using it. this is a replacement.
+					InsertAction::Replace(Node::Extension(
+						existing_key.encoded_leftmost(cp, false),
+						self.storage.alloc(Stored::New(augmented_low)).into()
+					))
+				}
+			}
+			Node::Extension(encoded, child_branch) => {
+				let existing_key = NibbleSlice::from_encoded(&encoded).0;
+				let cp = partial.common_prefix(&existing_key);
+				if cp == 0 {
+					trace!(target: "trie", "no-common-prefix, not-both-empty (exist={:?}; new={:?}): TRANSMUTE,AUGMENT", existing_key.len(), partial.len());
+
+					// partial isn't empty: make a branch here
+					// extensions may not have empty partial keys.
+					assert!(!existing_key.is_empty());
+					let idx = existing_key.at(0) as usize;
+
+					let mut children = empty_children();
+					children[idx] = if existing_key.len() == 1 {
+						// direct extension, just replace.
+						Some(child_branch)
+					} else {
+						// more work required after branching.
+						let ext = Node::Extension(existing_key.mid(1).encoded(false), child_branch);
+						Some(self.storage.alloc(Stored::New(ext)).into())
+					};
+
+					// continue inserting.
+					InsertAction::Replace(self.insert_inspector(Node::Branch(children, None), partial, value).unwrap_node())
+				} else if cp == existing_key.len() {
+					trace!(target: "trie", "complete-prefix (cp={:?}): AUGMENT-AT-END", cp);
+
+					// fully-shared prefix.
+
+					// insert into the child node.
+					let (new_child, changed) = self.insert_at(child_branch, partial.mid(cp), value);
+					let new_ext = Node::Extension(existing_key.encoded(false), new_child.into());
+					if changed {
```

Comment:
it would reduce code size itself, but the semantics here are the important part. this is why i include comments at each such site.

adding a function like this could make it too easy to do the wrong thing in situations where you should ignore the `changed` flag